### PR TITLE
Realistic repeats

### PR DESCRIPTION
**Pull request by robertjmoore**
Addresses several issues:
* https://github.com/RJMetrics/magenerator/issues/30 (create realistic distribution of order status values)
* https://github.com/RJMetrics/magenerator/issues/25 (add UTM tags to the sales_flat_order table showing a realistic distribution of customer source, medium, campaign values)
* https://github.com/RJMetrics/magenerator/issues/14 (add a base_discount_amount field and values to sales_order_flat; only have nonzero values in cases where coupon codes appear)
* https://github.com/RJMetrics/magenerator/issues/24 (the big one: alter the behavior of the data generation to create a distribution of purchases that creates repeat purchase probabilities that increase in likelihood with each purchase)
* also fixed an issue where it was possible for a users created_at date to be later than the created_at date of orders they had placed
* also updated README to include missing npm install of "should"

Tested by generating data and loading into mysql and running various queries to test desired behavior; 

Sample query to test repeat probability distribution:
```
select count(*) as customers, purchaseCount from (select customer_id, count(*) as purchaseCount from sales_flat_order group by customer_id) a group by purchaseCount;
```

### CODE REVIEW

File: generate.coffee [hunk 1]
```diff
@@ -94,14 +94,43 @@ generateAddresses = (total) ->
...
-    customer = getRandomItem(customers)
     address = getRandomItem(addresses)
     couponCode = if chance.bool({likelihood: 10}) then getRandomItem(COUPONS) else null
     items = getItems(products, ITEMS_MIN, ITEMS_MAX)
-    createdAt = getRandomDate()
     grandTotal = getCartValue(items)
     shippingAmount = getRandomItem([1.99,3.99,6.99])
+    discountAmount = 0.00
+    if(couponCode)
+      discountAmount = -1*(getRandomItem([0.05,0.1,0.25])*grandTotal).toFixed(2)
```
Comment (bengarvey) [inline generate.coffee]: `-1*(getRandomItem([0.05,0.1,0.25])*grandTotal).toFixed(2)` is a lot of logic and should be moved to a sub function.

File: generate.coffee [hunk 2]
```diff
@@ -118,11 +147,30 @@ generateOrders = (total, customers, addresses, products) ->
...
+      utm_source: utmSource
+      utm_medium: utmMedium
+      utm_campaign: utmCampaign
       created_at: createdAt
       updated_at: createdAt
     orders.push(order)
+    orderCounts[customer.entity_id] = if orderCounts[customer.entity_id] then orderCounts[customer.entity_id] + 1 else 1
   return orders
 
+getCustomerToBuy = (customers, orderCounts) ->
```
Comment (bengarvey) [inline generate.coffee]: This is pretty slick. Instead of getCustomerToBuy let's name this getCustomerToBuyFavoringRepeats

File: generate.coffee [hunk 3]
```diff
@@ -161,7 +209,12 @@ getCartValue = (items) ->
   return total.toFixed(2)
 
 getOrderStatus = () ->
-  return "complete"
+  if chance.bool({likelihood: 83})
```
Comment (bengarvey) [inline generate.coffee]: In coffeescript you can make this a little simpler by writing it like this

```
if chance.bool({likelihood: 83) then "complete"
else if chance.bool({likeliood: 90}) then "processing"
else getRandomItem(['pending','canceled','picked','shipped','picking']) 
```

File: generate.coffee (hunk 2 above)
Comment (bengarvey) [inline generate.coffee]: Were you able to verify that this algorithm creates the proper incrementing event probability?

File: generate.coffee [hunk 4]
```diff
@@ -6,10 +6,10 @@ csvParse = require "csv-parse"
 async = require "async"
 require "should"
 
-TOTAL_CUSTOMERS = 20000
-TOTAL_ADDRESSES = 20000
+TOTAL_CUSTOMERS = 10000
+TOTAL_ADDRESSES = 10000
 TOTAL_PRODUCTS = 200
-TOTAL_ORDERS = 30000
+TOTAL_ORDERS = 30000 #keep ratio of orders to customers at least 3:1 to allow interesting repeat ratios to form
```
Comment (bengarvey) [inline generate.coffee]: Good comment.

**Comment by bengarvey** [comment]
@robertjmoore back to you

File: generate.coffee [hunk 5]
```diff
@@ -94,14 +94,43 @@ generateAddresses = (total) ->
...
   for index in [1..total]
-    customer = getRandomItem(customers)
     address = getRandomItem(addresses)
     couponCode = if chance.bool({likelihood: 10}) then getRandomItem(COUPONS) else null
     items = getItems(products, ITEMS_MIN, ITEMS_MAX)
-    createdAt = getRandomDate()
     grandTotal = getCartValue(items)
     shippingAmount = getRandomItem([1.99,3.99,6.99])
+    discountAmount = 0.00
+    if(couponCode)
```
Comment (robertjmoore) [inline generate.coffee]: done

File: generate.coffee (hunk 1 above)
Comment (robertjmoore) [inline generate.coffee]: done

File: generate.coffee [hunk 6]
```diff
@@ -94,14 +94,43 @@ generateAddresses = (total) ->
...
+      discountAmount = -1*(getRandomItem([0.05,0.1,0.25])*grandTotal).toFixed(2)
+    createdAt = getRandomDate()
+    customer = getCustomerToBuy(customers, orderCounts)
+    if customer.created_at > createdAt
+      customer.created_at = createdAt #make sure customer created at or before first order placed
+
+    #set UTM parameters, whose values are dependent upon each other
```
Comment (robertjmoore) [inline generate.coffee]: done

File: generate.coffee (hunk 2 above)
Comment (robertjmoore) [inline generate.coffee]: yes I ran the whole analysis (core query in my PR description) and it works beautifully

File: generate.coffee (hunk 3 above)
Comment (robertjmoore) [inline generate.coffee]: cool beans, done

**Comment by robertjmoore** [comment]
Back to you @bengarvey -- tested, loaded, data inspected for proper behavior

**Comment by bengarvey** [comment]
Approved!

**Pull request merged**